chainerx_cc/chainerx/routines/creation.cc

```diff
@@ -345,4 +347,119 @@ Array Linspace(
+
+    std::vector<Shape> shapes;
+    shapes.reserve(inputs.size());
+    std::transform(inputs.begin(), inputs.end(), std::back_inserter(shapes), [](const Array& array) { return array.shape(); });
+
+    BackwardBuilder bb{"meshgrid", in_refs, out_refs};
+    if (BackwardBuilder::Target bt = bb.CreateTarget()) {
+        bt.Define([shapes = std::move(shapes), kind, dtype = inputs[0].dtype(), &device = inputs[0].device()](BackwardContext& bctx) {
+            int64_t ndims = bctx.input_count();
+
+            auto get_axis = [ndims, kind](int64_t dim) {
+                // Handle "XY" case
+                if (kind == "xy" && dim == 0) {
+                    dim = 1;
+                } else if (kind == "xy" && dim == 1) {
+                    dim = 0;
+                }
+                Axes axes;
+                for (int i = 0; i < ndims; i++) {
+                    if (i == dim) {
+                        continue;
+                    }
+                    axes.push_back(i);
+                }
+                return axes;
+            };
+
+            for (size_t i = 0; i < bctx.output_count(); ++i) {
+                const nonstd::optional<Array>& gy = bctx.output_grad(i);
+                bctx.input_grad(i) = gy.has_value() ? (*gy).Sum(get_axis(i)) : Zeros(shapes[i], dtype, device);
+            }
+        });
+    }
+    bb.Finalize();
+}
+
+}  // namespace
+
+std::vector<Array> Meshgrid(const std::vector<Array>& arrays, const nonstd::optional<std::string>& indexing) {
+    Shape shape;
+    Shape broadcast_shape;
+    std::vector<Shape> arr_reshapes;
+    std::vector<Array> grid_arrs(arrays.size());
+
+    // Algo
+    //
+    // Step 1: Reshape/View each array as broadcastable based
+    // on number of input vectors.
+    // Eg. For tuple of vectors (n1, n2, n3)
+    // where ni is length of that vector.
+    // After this step for Vector 1 , we will reshape it as
+    // (n1, 1, 1) , Vector 2 as (1, n2, 1)
+    //
+    // Step 2: Broadcast each vector to the shape
+    // if (indexing == "ij") -> (n1, n2, n3)
+    // else if (indexing == "xy") -> (n2, n1, n3)
+    // Note : For "xy" only n1 and n2 swap their places
+    //        all others are same as "ij"
+
+    // Step 1
+    for (const auto& array : arrays) {
+        shape.push_back(1);
+        broadcast_shape.push_back(array.GetTotalSize());
+    }
+
+    // Shape for each array based on number of arrays.
+    for (unsigned int i = 0; i < arrays.size(); i++) {
+        Shape temp_shape(shape.begin(), shape.end());
+        temp_shape[i] = arrays[i].GetTotalSize();
+        arr_reshapes.push_back(temp_shape);
+    }
+
+    // Referred from numpy documentation and source.
+    std::string kind = indexing.value_or("xy");
+    if (kind == "xy") {
+        std::swap(arr_reshapes[0][0], arr_reshapes[0][1]);
+        std::swap(arr_reshapes[1][0], arr_reshapes[1][1]);
+        std::swap(broadcast_shape[0], broadcast_shape[1]);
+    } else if (kind == "ij") {
+    } else {
```

Comment:
Could you write a comment above this line, just to notify the reader that this empty branch is intentional?

Reply:
Refactored.